stepflow-rs/crates/stepflow-core/src/workflow/json_path.rs

```diff
@@ -0,0 +1,347 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements.
+// See the NOTICE file distributed with this work for additional information regarding copyright
+// ownership.  The ASF licenses this file to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance with the License.  You may obtain a
+// copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied.  See the License for the specific language governing permissions and limitations under
+// the License.
+
+use schemars::JsonSchema;
+use serde::de::{self, Visitor};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::fmt;
+
+/// A single part of a JSON path.
+#[derive(Debug, Clone, PartialEq, Hash, Eq)]
+pub enum PathPart {
+    /// Access a field by name
+    Field(String),
+    /// Access an array element by index
+    Index(usize),
+    /// Access an element by string key (for arrays that can be indexed by string)
+    IndexStr(String),
+}
+
+/// A JSON path represented as a sequence of path parts.
+/// This type serializes and deserializes as a string but internally
+/// maintains a structured representation for efficient processing.
+#[derive(Debug, Clone, PartialEq, Hash, Eq)]
+pub struct JsonPath(Vec<PathPart>);
+
+impl JsonPath {
+    /// Create a new empty path
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    /// Create a path from a vector of parts
+    pub fn from_parts(parts: Vec<PathPart>) -> Self {
+        Self(parts)
+    }
+
+    /// Get the parts of the path
+    pub fn parts(&self) -> &[PathPart] {
+        &self.0
+    }
+
+    /// Check if the path is empty
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn outer_field(&self) -> Option<&str> {
+        match self.0.first() {
+            Some(PathPart::Field(name)) => Some(name),
+            Some(PathPart::IndexStr(name)) => Some(name),
+            _ => None,
+        }
+    }
+
+    /// Parse a string path into a JsonPath
+    pub fn parse(s: &str) -> Result<Self, String> {
+        if s.is_empty() {
+            return Ok(Self::new());
+        }
+
+        let s = match s.trim().strip_prefix("$") {
```

Comment:
Maybe pull out common strings into constants? Is rust better with optimising naked literals?

Reply:
Rust is pretty good about this -- they'll go into a single thing in the generated code. I think I'll keep these because I find it easier to read `"$"` and think "that is a dollar sign" than read `DOLLAR_SIGN`.